src/v/serde/avro/parser.cc

```diff
@@ -0,0 +1,170 @@
+        }
+        case ::avro::AVRO_ARRAY: {
+            auto element_count = decode_element_count();
+            auto ret = std::make_unique<parsed::list>();
+
+            if (element_count > 0) {
+                ret->elements.reserve(element_count);
+                for (size_t i = 0; i < element_count; ++i) {
+                    ret->elements.push_back(
+                      co_await parse_node(node->leafAt(0), state));
+                }
+                // avro always writes 0 as the array footer, first element count
+                // field is skipped if array is empty
+                _parser.read_varlong();
+            }
+            co_return std::make_unique<parsed::message>(std::move(ret));
+        }
+        case ::avro::AVRO_MAP: {
+            auto element_count = decode_element_count();
+            auto ret = std::make_unique<parsed::map>();
+            if (element_count > 0) {
+                ret->entries.reserve(element_count);
+                for (size_t i = 0; i < element_count; ++i) {
+                    auto key = parse_string();
+                    ret->entries.emplace_back(
+                      std::move(key),
+                      co_await parse_node(node->leafAt(1), state));
+                }
+                // avro always writes 0 as the map footer, first element count
+                // field is skipped if the map is empty
+                _parser.read_varlong();
+            }
+            co_return std::make_unique<parsed::message>(std::move(ret));
+        }
+        case ::avro::AVRO_UNION: {
+            auto idx = parse_length();
+            auto union_branch = node->leafAt(idx);
+            co_return std::make_unique<parsed::message>(
+              std::make_unique<parsed::avro_union>(parsed::avro_union{
+                .branch = static_cast<size_t>(idx),
+                .message = co_await parse_node(union_branch, state)}));
+        }
+        case ::avro::AVRO_FIXED: {
+            // fixed type is special as it does not have the size included in
+            // the binary representation
+
+            co_return std::make_unique<parsed::message>(
+              _parser.share(node->fixedSize()));
+        }
+        case ::avro::AVRO_SYMBOLIC:
+            // this is a special case used when dealing with nested types, we
+            // need to resolve the original type and parse it
+            co_return co_await parse_node(::avro::resolveSymbol(node), state);
+        case ::avro::AVRO_UNKNOWN:
+            throw std::logic_error{
+              "serde avro parser encountered an unknown type"};
+        }
+    }
+
+    size_t parse_length() {
+        auto [sz, _] = _parser.read_varlong();
+        if (sz < 0) {
+            throw std::runtime_error(fmt::format(
+              "Invalid element length: {}, length can not be negative.", sz));
+        }
+        return sz;
+    }
+
+    iobuf parse_string() {
+        auto sz = parse_length();
+        return _parser.share(sz);
```

Comment:
I am OK with sharing (I forget what I do with protobuf), but let's add that to the `parse` documentation.

EDIT looks like we do the same for protobuf: https://github.com/redpanda-data/redpanda/blob/1754e471e487e66f568e5128b61d1261965cefe5/src/v/serde/protobuf/parser.h#L110-L111